DeviceManager/DeviceManager.py

```diff
@@ -94,12 +94,50 @@ def create_device():
     """ Creates and configures the given device (in json) """
     try:
         tenant = init_tenant_context(request, db)
-        device_data, json_payload = parse_payload(request, device_schema)
-        device_data['id'] = generate_device_id()
-        device_data.pop('templates') # handled separatly by parse_template_list
-        orm_device = Device(**device_data)
-        parse_template_list(json_payload['templates'], orm_device)
-        auto_create_template(json_payload, orm_device)
+        try:
+            count = int(request.args.get('count', '1'))
+            clength = len(str(count))
+        except Exception as e:
+            LOGGER.error(e)
+            raise HTTPRequestError(400, "If provided, count must be integer")
+        result = None
+        devices = []
+        for i in range(0, count):
+            device_data, json_payload = parse_payload(request, device_schema)
+            device_data['id'] = generate_device_id()
+            device_data['label'] = device_data['label'] + "_%0*d" % (clength, i)
+            device_data.pop('templates', None) # handled separatly by parse_template_list
+            orm_device = Device(**device_data)
+            parse_template_list(json_payload.get('templates', []), orm_device)
+            auto_create_template(json_payload, orm_device)
+            db.session.add(orm_device)
+
+            devices.append({'id': device_data['id'], 'label': device_data['label']})
+
+            full_device = serialize_full_device(orm_device)
+
+            # TODO remove this in favor of kafka as data broker....
+            ctx_broker_handler = OrionHandler(service=tenant)
+            ctx_broker_handler.create(full_device)
+
+            kafka_handler = KafkaHandler()
+            kafka_handler.create(full_device, meta={"service": tenant})
+
+        try:
+            db.session.commit()
+        except IntegrityError as error:
+            handle_consistency_exception(error)
+
+        if count == 1:
+            result = json.dumps({
+                'message': 'device created',
+                'device': full_device
+            })
+        else:
+            result = json.dumps({
+                'message': 'devices created',
+                'devices': devices
```

Comment:
For bulk creation, only the id and label of each device is returned, differently from a single device creation where the full device is returned?

Reply:
This return does not match the description on the apib where the entire devices are returned

Reply:
When I first wrote this, it seemed to me that, for very large bulk creations (i.e. hundreds of devices), returning full device descriptions would make for infeasible response lengths. Thus the idea of returning only the list of labels and ids for the created devices.

What I think may be a compromise between the two is perhaps to add a new query option to enforce full descriptions being returned, with a default to return only the "bulk" response (i.e. labels and ids)

Still, my proposal is to allow full descriptions to be returned only if a single device is created (or just a very small number (tens of devices).

For the API description, when rendered, the provided value for the querystring fields is appended to the URL, so on 200 status code I've used the result for that call:

![image](https://user-images.githubusercontent.com/25080094/34519944-b6ac7b3e-f06d-11e7-88dd-71bc1666b6ce.png)

Reply:
I understand the proposal for shorter response messages and agree with it.
Having that said, for someone that only looks at the API description for using the API it might be misleading if you are expecting a full device return and with a slightly different query with a higher count, the response is different.

Reply:
The new proposal (usage of verbose flag with device summary as default) was implemented on b8c2ea0